docs/documentation/code-format-lint.md

```diff
@@ -0,0 +1,98 @@
+# Linting and Code Formatting
+
+In our [packaging guide](https://www.pyopensci.org/python-package-guide/package-structure-code/code-style-linting-format.html), we
+discuss various linters and code formatters used in the scientific Python
+ecosystem, including tools like:
+
+* [black](https://github.com/psf/black)
+* [flake8](https://github.com/pycqa/flake8)
+* [isort](https://pycqa.github.io/isort/)
+
+[Ruff](https://github.com/charliermarsh/ruff) is a tool that has been
+quickly becoming more popular. It is written in the Rust programming language
+and is fast. We like Ruff because it can perform both code formatting
+(similar to Black) and linting (similar to Flake8). This makes your setup
+simpler.
+
+## Running Linters with pre-commit
+
+There are a few ways to call the linters when you are working on your code.
+
+One option is to use pre-commit hooks. [pre-commit](https://pre-commit.com/)
+runs any defined linters, code and text formatters, spellcheckers, and other
+tools on your code locally when you use `git commit` to make a change. For
+example:
+
+```bash
+git commit -m "message here"
+```
+
+By configuring pre-commit hooks, you can automatically run tools like Ruff and
+codespell on your code and documentation every time you commit. This ensures
+consistency and catches errors early.
+
+While pre-commit is a powerful tool to add to your workflow, many do not like it:
+
+* For beginner contributors, running pre-commit hooks can be confusing. You need
+  to understand that each time you commit, it will run the checks. If it finds
+  issues in your files, it will not actually commit your files to history. This
+  can be confusing for even seasoned developers if they haven't used pre-commit
+  before.
+* Some prefer to set up autoformatters that run every time you save a file in
+  your preferred code editor (like VSCode). More on that below.
+
+## Pre-commit.ci Bot
+
+The [pre-commit CI bot](https://pre-commit.ci/) integrates with your GitHub
+repository to run pre-commit checks in an online continuous integration pipeline.
+The bot will run any pre-commit hooks that you have set up on new pull requests
+submitted to your repository.
+
+Pre-commit.ci can be a nice tool to use for pull requests if set up correctly.
+Ideally, you can set pre-commit CI to run only when you call it in a pull
+request. This way, if you have a new contributor (or a seasoned one) who doesn't
+want to set up pre-commit locally, or someone who wants to submit a pull request
+(e.g., as a first contribution!) fully from the GitHub interface, you can enable
+the bot to run pre-commit checks and fixes in the pull request yourself, as a
+maintainer, before you merge the PR.
```

Comment:
One thing to bear in mind with this is that when the bot pushes a fix, contributors need to pull down that fix before making further changes, which can be confusing.

Reply:
oh right - good point! Would being a bit stronger about the language - only run pre-commit right before you merge the PR help clarify?

Reply:
I meant if the bot itself pushes a fix it can be confusing, which I don't think can be configured to run on demand.

Reply:
AHHH ok. i do use the bot on demand - even in our repos here at pyos.

you can set the bot up to NOT autoformat and just check the repo. then when ready, use `pre-commit autofix` to call the bot. 

my preferred workflow is

setup precommit.
Allow users to use it (or not)
i use it locally as a maintainer,
when a contributor submits a pr, i review and ignore linting until the end. when we are ready to merge, run `precommit autofix` on the pr.  and the bot does it's thing.

the bot can't fix spelling errors. 

what i'm trying to figure out here that i wonder if you / jupyter has solved is how should we suggest people run linters if pre-commit is confusing. (i understand it is confusing!)

Reply:
Please remember also the security issues that @ofek mentioned on Slack. I think, that's important to note here.